### PR TITLE
Add qmake support

### PR DESCRIPTION
**Pull request by kimmoal**
Some feedback/resolution needed:
- Points parsing: using copy paste code from the tmc-langs/regex branch, this should be merged somehow. We need to parse the function "task" name so it needs to be overrideable.

Still unfinished functionality AKA **TODO**
- [x] QmakeStudentFilePolicy: user files in static
- [x] POINT defines for all test cases
- [x] Final template / exercise structure
- [x] More tests for the main plugin, exercise scanning.
- [x] More tests.
- [x] More documentation
- [x] Currently only tested on Ubuntu. Windows/macOS needs testing.
- [x] Testing for handling a non-Qt project (eg. makefile)
- [ ] look into exclude thing from maven-surefire-plugin in pomxml
- [x] remove parseTestCases from constructor

### CODE REVIEW

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
+package fi.helsinki.cs.tmc.langs.qmake;
+
+import fi.helsinki.cs.tmc.langs.domain.RunResult;
+import fi.helsinki.cs.tmc.langs.domain.TestResult;
+import fi.helsinki.cs.tmc.langs.utils.TestUtils;
+import java.io.IOException;
```
Comment (ljleppan) [inline tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java]: Formatting

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        Path testPath = null;
+        try {
+            testPath = constructTestOutput(testCases);
+        } catch (Exception e) {
+            fail("Error creating or parsing output file: " + e.getMessage());
+        }
+        return new QTestResultParser(testPath);
+    }
+
+    private Path constructTestOutput(List<QTestCase> testCases) throws IOException {
```
Comment (ljleppan) [inline tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java]: I don't fully understand what is happening here. Why are we testing against this generated output rather than real output?

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QmakePluginTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
+package fi.helsinki.cs.tmc.langs.qmake;
+
+import com.google.common.base.Optional;
```
Comment (ljleppan) [inline tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QmakePluginTest.java]: Import ordering

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QmakePluginTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        assertTrue(isExerciseTypeCorrect("passing_multiple_lib"));
+        assertTrue(isExerciseTypeCorrect("passing_multiple_lib_same_point"));
+
+        assertTrue(isExerciseTypeCorrect("failing_single_lib_not_compiling"));
+        assertTrue(isExerciseTypeCorrect("failing_nolib"));
+        assertTrue(isExerciseTypeCorrect("failing_nolib_same_point"));
+        assertTrue(isExerciseTypeCorrect("failing_single_lib"));
+        assertTrue(isExerciseTypeCorrect("failing_single_lib_not_compiling"));
+        assertTrue(isExerciseTypeCorrect("failing_single_lib_same_point"));
+        assertFalse(isExerciseTypeCorrect(""));
```
Comment (ljleppan) [inline tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QmakePluginTest.java]: Should this also check against a project that is otherwise valid but just not a QT project?

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    public String getPluginName() {
+        return "qmake";
+    }
+
+    /**
+     * Resolve the exercise .pro file from exercise directory. The file should
+     * be named after the directory.
+     */
+    private Path getProFile(Path basePath) {
+        Path proFile = new File(basePath.toFile().getName() + ".pro").toPath();
```
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Thanks, did not know about the old/new java.io API!

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java (hunk 2 above)
Comment (kimmoal) [reply tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java]: The test output XML files could be made static in the test resources, but every different edge case would have to have its own file. Inspiration to generate the test output came from the MakePlugin tests, where, I guess, those tests use the same reasoning.

Of course these generated test cases do not really test against the **true** output of the test framework, but the output is general enough (IMHO) to warrant generative output!

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                    tests.get(testName).add(point);
+                }
+            }
+            for (String testName : tests.keySet()) {
+                List<String> points = tests.getOrDefault(testName, new ArrayList<String>());
+                descs.add(new TestDesc(testName, ImmutableList.<String>copyOf(points)));
+            }
+            return Optional.of(ImmutableList.copyOf(descs));
+        } catch (IOException e) {
+            // We could scan the exercise but that could be a security risk
```
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: This is copy paste from tmc-langs/regex branch, so I do not know what this code exactly does, @nygrenh can probably comment on this. My best guess is: scanning for student generated files could DoS the regex parser?

[the remark this answers is not in the document]
Comment (ljleppan) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: I believe the original code you are referencing had the potential of running (or maybe compiling?) some student-provided Java code outside of the TMC sandbox. Check wit @nygrenh.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ljleppan) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Most of the old `io` package is fine, but it has enough stupid "gotcha" stuff that it's just easier to avoid it wholesale :)

For example, if you have two `java.io.URL` objects, [`java.io.URL.equals()`](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#equals(java.lang.Object)) ends up making two DNS calls and thinks the hosts are equal if they resolve to the same IP: something that might have been only a slight horror in the age of dedicated self-hosted servers, but is completely useless and actively harmful in the age of cloud computing.

File: tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java (hunk 2 above)
Comment (ljleppan) [reply tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java]: I'm mostly concerned of a scenario where this code ends up generating output that is out-of-sync with the output of the real system. If that's not a concern, it's probably fine.

Comment (ljleppan) [reply tmc-langs-qmake/src/test/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParserTest.java]: Also, check out [`try`-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) to ensure we close the file handle in case of an exception.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    private static final String DOC_NULL_ERROR_MESSAGE = "Failed to parse test results";
+    private static final String SAX_PARSER_ERROR = "SAX parser error occured";
+    private static final String PARSING_DONE_MESSAGE = "Qt test cases parsed.";
+
+    private static final Logger log = LoggerFactory.getLogger(QTestResultParser.class);
+
+    private final List<TestResult> tests;
+
+    public QTestResultParser(Path testResults) {
+        this.tests = parseTestCases(testResults);
```
Comment (nygrenh) [inline tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java]: Reading a file in constructor unexpected.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+import java.util.regex.Pattern;
+
+public final class QmakePlugin extends AbstractLanguagePlugin {
+
+    private static final Path TMC_TEST_RESULTS = Paths.get("tmc_test_results.xml");
+
+    // POINT(exercise_name, 1)
+    private static final Pattern POINT_PATTERN
+            = Pattern.compile("POINT\\(\\s*(\\w+),\\s*(\\w+)\\s*\\)\\s*;");
+    private static final Pattern COMMENT_PATTERN
```
Comment (nygrenh) [inline tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Comment that this matches commented code so that it can be deleted before getting the points with the points pattern.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+import java.util.regex.MatchResult;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class QmakePlugin extends AbstractLanguagePlugin {
+
+    private static final Path TMC_TEST_RESULTS = Paths.get("tmc_test_results.xml");
+
+    // POINT(exercise_name, 1)
+    private static final Pattern POINT_PATTERN
```
Comment (nygrenh) [inline tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Regex that matches all points, and puts the point name into the first match group. Also explain the other match groups.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+        }
+    }
+
+    private Optional<ProcessResult> run(String[] command, Path dir) {
+        ProcessRunner runner = new ProcessRunner(command, dir);
+
+        try {
+            return Optional.of(runner.call());
+        } catch (Exception e) {
```
Comment (nygrenh) [inline tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: List specific exceptions

**Comment by nygrenh** [comment]
What's the status of this pull request? You still have a lot of checkboxes unchecked in the description.

File: tmc-langs-qmake/pom.xml [hunk 11]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>fi.helsinki.cs.tmc</groupId>
+        <artifactId>tmc-langs</artifactId>
+        <version>0.7.7-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/pom.xml]: Apparently it is not supported?

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+            // Convert node to element, messages.item(i) returns a node
+            Element message = (Element) messages.item(i);
+            Element desc = (Element) message.getElementsByTagName("Description").item(0);
+            String text = desc.getTextContent();
+            if (text.matches("^(TMC:.*)")) {
+                String[] split = text.split("\\.");
+                String result = split[1];
+                for (int j = 2; j < split.length; j++) {
+                    result += "." + split[j];
+                }
```
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java]: This is basically skipping the name of the test function and using the rest as the point which could contain a ".", not replacing anything.

I also agree on the 1.8 part.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+
+        try {
+            ProcessResult qmakeBuild = buildWithQmake(shadowDir);
+            if (qmakeBuild.statusCode != 0) {
+                log.error("Building project with qmake failed: {}", qmakeBuild.errorOutput);
+                return filledFailure(qmakeBuild.errorOutput);
+            }
+        } catch (Exception e) {
+            log.error("Building project with qmake failed", e);
+            throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: What do you mean by custom (checked) exception? Like `QmakeBuildException` and `MakeBuildException` etc? Would it be helpful to include stdout/stderr output to exceptions from ProcessResult?

[the remark this answers is not in the document]
Comment (ljleppan) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Yeah, something like that. For example the Java plugins throw a [`fi.helsinki.cs.tmc.langs.java.exception.TestRunnerException `](https://github.com/testmycode/tmc-langs/blob/master/tmc-langs-java/src/main/java/fi/helsinki/cs/tmc/langs/java/exception/TestRunnerException.java) in a somewhat analogous situation.

I don't know how familiar you are with checked vs unchecked exceptions, but here's a short primer that also represents a pretty sensible stance on when to use unchecked exceptions: https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html.  

Unchecked exceptions (`RuntimeException`, `Error` and their subclasses) are a bit controversial and Google will find a plethora of extreme opinions. The standard stance seems to be that checked exceptions are used when the exception is such that the code calling your code can be expected to recover and unchecked exceptions are used for behavior that is clearly a bug in the code calling your code.

As a concrete example, if you call a method that excepts an instance of `Thing` as it's parameter with `null` argument, it's OK for that method to throw a `NullPointException`, which is unchecked and not part of the method signature, because *your code* is using the method wrong. In other words, the bug is not in the method but the code calling it. Similarly it's fine for `divide(0, 0)` to throw an unchecked `ArithmeticException` since clearly the code calling the function is doing it wrong.

At the same time, we want to throw a checked exception like `IOException` when a disk read fails, as that's something that the calling code can be expected to -- or atleast attempt to -- recover from.

If that made no sense, hit me up on Slack :)

---

As for including the stdout/stderr, it's probably unnecessary to include them in the exception itself, but logging them is probably a sensible thing to do. At least log the stderr when things go wrong. Essentially consider this: what information would *you* want to have when you debug a problem happening on someone else's computer and you only have the log files? :)

File: tmc-langs-qmake/pom.xml [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    </parent>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.compiler.source>1.7</maven.compiler.source>
+        <maven.compiler.target>1.7</maven.compiler.target>
```
Comment (nygrenh) [inline tmc-langs-qmake/pom.xml]: In 55e4dfb14582ad6c933d85a6ea3e35116148398b we dropped support for Java 7 so these lines should be removed so that the maven compiler inherits from the parent pom.

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (nygrenh) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java]: We have already ditched Java 1.7 on master.

[the remark this answers is not in the document]
Comment (ljleppan) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QTestResultParser.java]: Sorry for misunderstanding the code the first time, I blame lack of sleep :) Looking at this in more detail, am I correct in my reading that this can be effectively simplified to `String result = text.substring(text.indexOf(".") +1);`?

File: tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kimmoal) [reply tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java]: Thank you for the lengthy explanation, tried searching for the concept and found the controversy (that was really the problem here). Will change unchecked exceptions to checked ones wherever they seem to be desired.

Should be logging stderr currently at every point when something goes wrong.
